library/ssl_tls13_generic.c

```diff
@@ -434,13 +472,66 @@ int mbedtls_ssl_tls13_process_certificate_verify( mbedtls_ssl_context *ssl )
 
 /*
  *
- * STATE HANDLING: Incoming Certificate, client-side only currently.
+ * STATE HANDLING: Incoming Certificate.
  *
  */
 
-/*
- * Implementation
+/* Coordination: Check if a certificate is expected.
+ * Returns a negative error code on failure, and otherwise
+ * SSL_CERTIFICATE_EXPECTED or
+ * SSL_CERTIFICATE_SKIP
+ * indicating whether a Certificate message is expected or not.
  */
+#define SSL_CERTIFICATE_EXPECTED   0
+#define SSL_CERTIFICATE_SKIP       1
+
+static int ssl_tls13_read_certificate_coordinate( mbedtls_ssl_context *ssl )
```

Comment:
Compared to the prototype in the upstream we have gone the way of, as far as possible, jumping certificate related states if there is nothing to do in the states. This avoids the duplication of checks and you can see the result here with all those checks that have already been done before. Here I think we could aim to get completely rid of this coordination function. One possible way is to note down in the handshake context if we sent a CertificateRequest message and then at the end of server Finished writing to jump directly to the client Finished processing state if we did not send a CertificateRequest message. The switch to the handshake transform for inbound traffic have to be moved somewhere else, probably at the end of server Finished writing for the time being. We will have to revisit that when we add support for early data but that's ok I think.

Reply:
Seems we need take the server finished PR https://github.com/Mbed-TLS/mbedtls/pull/5749 as the preciding one.
Cause we have no psk currently, the new field in handshake maybe similar with the authmode.

Reply:
If easier, yes please rebase on top of #5749.